### PR TITLE
fixed broken leumi card adapter

### PR DESCRIPTION
**Pull request by liranso**
LeumiCard changed their implementation - they now use REST calls for each search in transactions page.
The change uses puppeteer to get into leumi card website, and then use REST to get the transaction data for each month (it was scraping the UI before, which now does not exists).

### CODE REVIEW

File: src/scrapers/leumi-card.js [hunk 1]
```diff
@@ -115,158 +90,50 @@ function getInstallmentsInfo(comments) {
...
+    type: getTransactionType(rawTransaction.planName),
+    date: moment(rawTransaction.purchaseDate).toISOString(),
+    processedDate: moment(rawTransaction.paymentDate).toISOString(),
+    originalAmount: -rawTransaction.originalAmount,
+    originalCurrency: rawTransaction.originalCurrency,
+    chargedAmount: -rawTransaction.actualPaymentAmount,
+    description: rawTransaction.merchantName.trim(),
+    memo: rawTransaction.comments,
+    installments: getInstallmentsInfo(rawTransaction.comments),
+    status: TRANSACTION_STATUS.COMPLETED,
```
[the remark this answers is not in the document]
Comment (liranso) [reply src/scrapers/leumi-card.js]: @esakal 
is it a proposal of a new behaviour or an old one that i did not support?

[the remark this answers is not in the document]
Comment (esakal) [reply src/scrapers/leumi-card.js]: It's a new feature of LeumiCard,I don't think they returned pending transactions in the old API. 

Now that they introduce such a use-case we must support so it is not exactly a regression nor a proposal :)

[the remark this answers is not in the document]
Comment (liranso) [reply src/scrapers/leumi-card.js]: sure :)
can you explain about the the new logic you proposed?
the screenshot you added shows that "paymentDate" is null, but you suggest checking "processedDate" value

[the remark this answers is not in the document]
Comment (esakal) [reply src/scrapers/leumi-card.js]: How about:
```
  const isPending = rawTransaction.paymentDate === null;
  return {
    processedDate: moment(isPending ? rawTransaction.purchaseDate :
      rawTransaction.paymentDate).toISOString(),
    status: isPending ? TRANSACTION_STATUS.PENDING : TRANSACTION_STATUS.COMPLETED,
  };
```
I don't think that we can leave the `processedDate` property value `null` and it make sense to use the same date of the purchase as this transaction in PENDING mode.

[the remark this answers is not in the document]
Comment (liranso) [reply src/scrapers/leumi-card.js]: Done
Take a look

**Comment by esakal** [comment]
@eshaham  indeed no need to fetch the rest content by page navigation. 
 @liranso can you adjust the code  per @eshaham  request?  
0.`import { fetchGetWithinPage } from '../helpers/fetch';`
1. change method `fetchTransactionsForMonth` as shown below
```
async function fetchTransactionsForMonth(page, navigateToFunc, monthMoment) {
  const url = getTransactionsUrl(monthMoment);
  const data = await fetchGetWithinPage(page, url);

  const transactionsByAccount = {};
  data.result.transactions.forEach((transaction) => {
    if (!transactionsByAccount[transaction.shortCardNumber]) {
      transactionsByAccount[transaction.shortCardNumber] = [];
    }

    const mappedTransaction = mapTransaction(transaction);
    transactionsByAccount[transaction.shortCardNumber].push(mappedTransaction);
  });

  return transactionsByAccount;
}
```
2. update `fetchTransactions` signature and body to pass page instead of browser to `fetchTransactionsForMonth`
3. remove method `getCurrentTransactions` which is not in use anymore

wdyt?

File: src/scrapers/leumi-card.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (erikash) [reply src/scrapers/leumi-card.js]: @esakal are you sure it doesn't make sense to leave it empty? putting the `purchaseDate` into the `processedDate` will break the `processedDate` semantics (as the transaction wasn't actually processed)

**Comment by liranso** [comment]
@esakal when i'm using "fetchGetWithinPage" i'm getting "Evaluation failed: [object Object]"
for some reason i'm not able to understand the cause of the error
i'm debugging with webstorm and defined the following node parameters - "-r babel-register"
i'm guessing it's somehow related to babel 

`
const {createScraper} = require('./src');

const date = new Date();
date.setMonth(date.getMonth() - 1);

(async function() {
    const scraper = createScraper({
        companyId: 'leumiCard', // mandatory; one of 'hapoalim', 'leumi', 'discount', 'otsarHahayal', 'visaCal', 'leumiCard', 'isracard', 'amex'
        startDate: date, // the date to fetch transactions from (can't be before the minimum allowed time difference for the scraper)
        combineInstallments: false, // if set to true, all installment transactions will be combine into the first one
        showBrowser: true, // shows the browser while scraping, good for debugging (default false)
        verbose: true, // include more debug info about in the output
        // browser : Browser // optional option from init puppeteer browser instance outside the libary scope. you can get browser diretly from puppeteer via `puppeteer.launch()` command.
    });

    const result = await scraper.scrape({
            username: 'XYZ',
            password: 'XYZ'
        }
    )
})();`

do you have an idea?

File: src/scrapers/leumi-card.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (esakal) [reply src/scrapers/leumi-card.js]: @erikash, I tried to keep the same behavior as other scrapers that expose pending transactions (leumi, discount). They seems to set a date also for pending transactions. I think people expect this field to be required and their code might break at runtime once we introduce an optional as `null` value. 

I think the term `Pending` is a bit misleading, as it doesn't say if its data is final and the transaction pending approval (either will remain or will be removed) or if it means that the transaction data is not final as well and might be modified once approved.

That being said, I don't might setting this value as `null` but in that case we should mention it in the `readme.md` file and consider upgrading major version to indicate breaking changes. @eshaham please let us know how you want to handle this one.

**Comment by esakal** [comment]
> esakal when i'm using "fetchGetWithinPage" i'm getting "Evaluation failed: [object Object]"
for some reason i'm not able to understand the cause of the error
i'm debugging with webstorm and defined the following node parameters - "-r babel-register"
i'm guessing it's somehow related to babel

I created a workable scraper. just copy&paste from [this gist](https://gist.github.com/esakal/75a291267530fa7a7980577ae5dd055c) into your PR and see if it works for you. If it works just update your PR accordingly. 

I didn't use the date filters as it seems to break the installments expected results. I created a [gist](https://gist.github.com/esakal/dc23700a5f0689ae47291b98c9d424d2) with that version but ended up reverting and using the monthly scrapping concept to get the expected installments behavior.

What I changed:
[x] use `fetchGetWithinPage` to fetch transactions
[x] in `fetchTransactions` filter months to include next month (to fetch all transactions known till today even if part of next month payment).
[x] removed unused methods used in previous implementation.

File: src/scrapers/leumi-card.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (erikash) [reply src/scrapers/leumi-card.js]: @esakal AFAIK `pending charges` amounts can be changed.. they are used to "hold" funds in cases that the final amount is not yet known (like an open tab in a hotel) or when buying gas in some gas stations... but i'm not sure about the semantics in Leumi Card :(

I liked your idea!
Maybe we can do both:
1. Keep the contract and publish a minor
2. In the next major change the default behavior.

@eshaham please advise...

**Comment by liranso** [comment]
@esakal 
Changed the PR
regarding this line
`const allMonths = getAllMonthMoments(startMoment, true);`
why do we need the next month?

File: src/scrapers/leumi-card.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (eshaham) [reply src/scrapers/leumi-card.js]: > 1. Keep the contract and publish a minor
> 2. In the next major change the default behavior.

I agree, let's not make a big change over this PR - I assume people are waiting for this scraper to be fixed ASAP, so let's get it out.

**Comment by esakal** [comment]
> @esakal
> Changed the PR
> regarding this line
> `const allMonths = getAllMonthMoments(startMoment, true);`
> why do we need the next month?

Great, I will test it and hopefully merge it today. Did it work for you with `getAllMonthMoments(startMoment, true);` set to `true`?

To answer your question. This library scrape all transactions created till today (provided start time from user).If you set this argument to `false` it will not scrape next payment transactions. Let say your payment date is the 5th of each month, if you set it to `false` you will not get any transaction between 6/07 till today. You can easily check this out by setting this value to `false` if your card payment date is less then the 15th of each month (as today is 15/07). Hope this clarify the reason.

**Comment by esakal** [comment]
@liranso it is such a small world - it was nice seeing you :)
I confirmed that it works and will merge it in the following minutes. Thanks a lot for your contribution!

**Comment by brafdlog** [comment]
Thanks!

**Comment by esakal** [comment]
@eshaham I merge the scraper into master, please deploy a new version we you can. Thanks!

**Comment by liranso** [comment]
@esakal :)

**Comment by brafdlog** [comment]
@eshaham can you publish a new version to npm?

**Comment by eshaham** [comment]
@brafdlog @esakal @liranso done 👍 
thanks for this great PR @liranso

**Pull request merged**